js/coinbase.js

```diff
@@ -292,14 +310,114 @@ module.exports = class coinbase extends Exchange {
         //           },
         //           "payout_at": "2015-02-18T16:54:00-08:00"
         //         }
+        //
+        //  SELL
```

Comment:
Not really sure how a SELL or BUY fits into parseTransaction, if it involves two currencies, then it's a trade, not a transaction? Can you elaborate on this part a bit?

Reply:
The coinbase `transactions` endpoint is very weird: 
https://developers.coinbase.com/api/v2#transactions

From their docs:
- `send` - Sent bitcoin/bitcoin cash/litecoin/ethereum to a bitcoin/bitcoin cash/litecoin/ethereum address or email (documentation)
- `request` - Requested bitcoin/bitcoin cash/litecoin/ethereum from a user or email (documentation)
- `transfer` - Transfered funds between two of a user’s accounts (documentation)
- `buy` - Bought bitcoin, bitcoin cash, litecoin or ethereum (documentation)
- `sell` - Sold bitcoin, bitcoin cash, litecoin or ethereum (documentation)
- `fiat_deposit` - Deposited funds into a fiat account from a financial institution (documentation)
-`fiat_withdrawal` - Withdrew funds from a fiat account (documentation)
- `exchange_deposit` - Deposited money into GDAX
- `exchange_withdrawal` - Withdrew money from GDAX
- `vault_withdrawal` - Withdrew funds from a vault account
- More to be added soon

It has everything. But the datasets show up differently

When you send bitcoin:
```
{
  "id": "57ffb4ae-0c59-5430-bcd3-3f98f797a66c",
  "type": "send",
  "status": "completed",
  "amount": {
    "amount": "-0.00100000",
    "currency": "BTC"
  },
  "native_amount": {
    "amount": "-0.01",
    "currency": "USD"
  },
  "description": null,
  "created_at": "2015-03-11T13:13:35-07:00",
  "updated_at": "2015-03-26T15:55:43-07:00",
  "resource": "transaction",
  "resource_path": "/v2/accounts/2bbf394c-193b-5b2a-9155-3b4732659ede/transactions/57ffb4ae-0c59-5430-bcd3-3f98f797a66c",
  "network": {
    "status": "off_blockchain",
    "name": "bitcoin"
  },
  "to": {
    "id": "a6b4c2df-a62c-5d68-822a-dd4e2102e703",
    "resource": "user",
    "resource_path": "/v2/users/a6b4c2df-a62c-5d68-822a-dd4e2102e703"
  },
  "instant_exchange": false,
  "details": {
    "title": "Sent bitcoin",
    "subtitle": "to User 2"
  }
}
```

When you buy bitcoin (the buy from the `transaction` endpoint gives you different data from the `buys` endpoint 🙃)
```
{
  "id": "8250fe29-f5ef-5fc5-8302-0fbacf6be51e",
  "type": "buy",
  "status": "pending",
  "amount": {
    "amount": "1.00000000",
    "currency": "BTC"
  },
  "native_amount": {
    "amount": "10.00",
    "currency": "USD"
  },
  "description": null,
  "created_at": "2015-03-26T13:42:00-07:00",
  "updated_at": "2015-03-26T15:55:45-07:00",
  "resource": "transaction",
  "resource_path": "/v2/accounts/2bbf394c-193b-5b2a-9155-3b4732659ede/transactions/8250fe29-f5ef-5fc5-8302-0fbacf6be51e",
  "buy": {
    "id": "5c8216e7-318a-50a5-91aa-2f2cfddfdaab",
    "resource": "buy",
    "resource_path": "/v2/accounts/2bbf394c-193b-5b2a-9155-3b4732659ede/buys/5c8216e7-318a-50a5-91aa-2f2cfddfdaab"
  },
  "instant_exchange": false,
  "details": {
    "title": "Bought bitcoin",
    "subtitle": "using Capital One Bank"
  }
}
```

Reply:
One option is i could filter buys and sells out, but that seems weird

Reply:
> The coinbase transactions endpoint is very weird:

They should've called it `ledger` I think... Because, as you know, there's a commonality between transactions and trades – they all change the balance. Some exchanges do have a separate endpoint for "all types of balance-changing entries", let's call it a *ledger* endpoint. I think, Kraken has it, and Bitstamp has it as well, Huobi and a few others probably. So, it's an endpoint that outputs a list of mixed-type entities, including trades and transactions (could be other types of balance-changing entries, like transfers from/to subaccounts, etc).

I'd suggest that we think a bit on how this should be properly unwinded into a set of unified methods, maybe introduce a new one and implement it where available from the exchange, something like `fetchLedger` or `fetchLedgerEntries` or `fetchTradesAndTransactions` and we should probably account for exchanges other than Coinbase here to avoid problems with other exchanges later. Your thoughts are welcome )

Reply:
The question I have is how to unify all objects in the response... it seems like that’s a losing battle because of how varied the responses are. I guess the intention for this endpoint from the exchange is to fetch everything and then filter what you need or get a cursory view of all of your data and then query the specifics where needed. 

I do like fetchLedger as it sticks with one call per endpoint and works semantically for this scenario.

Reply:
> I do like fetchLedger as it sticks with one call per endpoint and works semantically for this scenario.

Yep, I suggest that we proceed in this direction. One other thing that we should think of is the signature of `fetchLedger`, as you may have noticed already, most methods that return arrays, have `since`, `limit` and `params` in their arguments, and the only difference between `fetchTransactions` and `fetchTrades` is the first argument. With trades it's a pair of currencies, that is *a symbol*, and with with transactions that is most often a single *currency code* (also often called *a symbol*, but not to be confused with a trading pair).

To me it could look like 

```
parseLedger (objects, since, limit) {
    // pseudocode
    let result = [];
    for (let i = 0; i < objects.length; i++) {
        let entry = objects[i];
        let type = this.safeString (entry, 'type');
        // simplified here, but should be buy/sell/etc for Coinbase
        if (type === 'trade') { 
            result.push (parseTrade (entry));
        } else if (type === 'transaction') {
            result.push (parseTransaction (entry));
        } else ...
    }
    return this.filterBySinceLimit (result, since, limit);
}

fetchLedger (symbol = undefined, since = undefined, limit = undefined, params = {}) {
    // here a symbol may relate to either a currency code or a trading pair symbol
    // pseudocode
    let request = {
        // ...
    }; 
    let response = await this.privateGetAccountsAccountIdTransactions (this.extend (request, params));
    return this.parseLedger (response, since, limit);
}
```

However, it may turn out to be more involved...